src/array.cr

```diff
@@ -1770,4 +1770,15 @@ class Array(T)
       end
     end
   end
+
+  # This method returns the i-th element. If i is equal to ary.size, it returns nil.
+  def bsearch(&block : T -> Int|Float|Bool?)
+    idx=bsearch_index(&block)
+    idx ? self[idx] : nil
```

Comment:
I am getting back into static typing and the one thing I don't like about general purpose functions returning nil is that there is a lot of nil handling or returning nil yourself.  I like the option of catching an error instead of having nil handling code around.  

If I understand correctly I think this is a good compromise.  You can incorporate nil if you want and with the overloads you have options where you don't have worry about nil.